File: lib/crowding/heuristic.ts
/**
 * Phase 1 rule-based bus fullness model.
 * Combines time, weather, and stop density into a 0–1 score and maps to crowd_risk + crowding.
 */

import type { WeatherFeatures } from "./weather"
import { getWeatherFeatures } from "./weather"
import { getTimeFeatures, getStopDensityFeature } from "./features"

export type CrowdLevel = "low" | "medium" | "high"

export type CrowdingResult = {
  crowd_risk: CrowdLevel
  crowding: { level: CrowdLevel; prob_full: number }
  /** Raw 0–1 score for debugging/tuning */
  score?: number
}

const DENSITY_WEIGHT: Record<string, number> = {
  low: 0.1,
  medium: 0.2,
  high: 0.35,
}

const THRESHOLD_LOW = 0.33
const THRESHOLD_HIGH = 0.66

function scoreToLevel(score: number): CrowdLevel {
  if (score <= THRESHOLD_LOW) return "low"
  if (score <= THRESHOLD_HIGH) return "medium"
  return "high"
}

/**
 * Compute crowding from features. Score in [0, 1], then map to levels.
 */
export function computeCrowdingFromFeatures(
  timeFeatures: ReturnType<typeof getTimeFeatures>,
  weather: WeatherFeatures,
  stopDensity: "high" | "medium" | "low"
): CrowdingResult {
  let score = 0.15 // base

  // Time: weekend = lower, rush/dismissal = higher
  if (timeFeatures.is_weekend) {
    score += 0.05
  } else {
    if (timeFeatures.in_rush_hour) score += 0.25
    if (timeFeatures.in_dismissal_window) score += 0.2
  }

  // Weather: precip or very cold → more riders
  if (weather.is_precipitating) score += 0.2
  if (weather.is_very_cold) score += 0.15

  // Stop density
  score += DENSITY_WEIGHT[stopDensity] ?? DENSITY_WEIGHT.medium

  score = Math.min(1, Math.max(0, score))

  const level = scoreToLevel(score)
  const prob_full = Math.round(score * score * 100) / 100 // slight curve

  return {
    crowd_risk: level,
    crowding: { level, prob_full },
    score,
  }
}

export type GetCrowdingOptions = {
  stopNameOrId?: string
  now?: Date
  weather?: WeatherFeatures | null
}

const DEFAULT_STOP = "University & Lake"

/**
 * Get crowding prediction for a stop. Fetches weather if not provided.
 */
export async function getCrowding(options: GetCrowdingOptions = {}): Promise<CrowdingResult> {
  const stop = options.stopNameOrId?.trim() || DEFAULT_STOP
  const now = options.now ?? new Date()

  const timeFeatures = getTimeFeatures(now)
  const stopDensity = getStopDensityFeature(stop)
  const weather = options.weather ?? (await getWeatherFeatures())

  return computeCrowdingFromFeatures(timeFeatures, weather, stopDensity)
}

File: backend/crowding/heuristic.py
"""
Phase 1 rule-based bus fullness model.
Combines time, weather, and stop density into a 0–1 score and maps to crowd_risk + crowding.
"""

from typing import Literal, Optional, TypedDict

from .weather import get_weather_features, WeatherFeatures
from .features import get_time_features, get_stop_density_feature

CrowdLevel = Literal["low", "medium", "high"]

DENSITY_WEIGHT: dict[str, float] = {
    "low": 0.1,
    "medium": 0.2,
    "high": 0.35,
}

THRESHOLD_LOW = 0.33
THRESHOLD_HIGH = 0.66

DEFAULT_STOP = "University & Lake"


def _score_to_level(score: float) -> CrowdLevel:
    if score <= THRESHOLD_LOW:
        return "low"
    if score <= THRESHOLD_HIGH:
        return "medium"
    return "high"


def compute_crowding_from_features(
    time_features: dict,
    weather: WeatherFeatures,
    stop_density: str,
) -> dict:
    score = 0.15

    if time_features.get("is_weekend"):
        score += 0.05
    if time_features.get("in_rush_hour"):
        score += 0.25
    if time_features.get("in_dismissal_window"):
        score += 0.2

    if weather.get("is_precipitating"):
        score += 0.2
    # is_very_cold: temp < 32°F (see weather.VERY_COLD_F)
    if weather.get("is_very_cold"):
        score += 0.15

    score += DENSITY_WEIGHT.get(stop_density, DENSITY_WEIGHT["medium"])
    score = max(0.0, min(1.0, score))

    level = _score_to_level(score)
    prob_full = round(score * score, 2)

    return {
        "crowd_risk": level,
        "crowding": {"level": level, "prob_full": prob_full},
        "score": score,
    }


def get_crowding(
    stop_name_or_id: Optional[str] = None,
    now=None,
    weather: Optional[WeatherFeatures] = None,
) -> dict:
    stop = (stop_name_or_id or "").strip() or DEFAULT_STOP
    time_features = get_time_features(now)
    stop_density = get_stop_density_feature(stop)
    weather_data = weather if weather is not None else get_weather_features()
    return compute_crowding_from_features(time_features, weather_data, stop_density)
